Contracts/Presale/PresaleContract.sol

```diff
@@ -23,88 +23,87 @@ contract PresaleContract is WithdrawContract {
                                 
     uint constant MINIMUM_PAYMENT_VALUE = 0.01 ether;//100 ether;
 
-    uint[] prices;
-    uint[] amounts; 
+    uint[] amountBoughtInOneEth;
+    uint[] amountsLeft; 
     
     uint totalRaisedEth = 0;
     uint soldTokensAmount = 0;
     address public mainVallet;
     function PresaleContract() public {
         mainVallet = owner;
-        prices.push(FIRST_PRESALE_PRICE);
-        prices.push(SECOND_PRESALE_PRICE);
-        prices.push(THIRD_PRESALE_PRICE);
+        amountBoughtInOneEth.push(FIRST_PRESALE_PRICE);
+        amountBoughtInOneEth.push(SECOND_PRESALE_PRICE);
+        amountBoughtInOneEth.push(THIRD_PRESALE_PRICE);
         
-        amounts.push(FIRST_PRESALE_AMOUNT * MINIMUM_TOKEN_VALUE_MULTIPLIER);
-        amounts.push(SECOND_PRESALE_AMOUNT * MINIMUM_TOKEN_VALUE_MULTIPLIER);
-        amounts.push(THIRD_PRESALE_AMOUNT * MINIMUM_TOKEN_VALUE_MULTIPLIER);
+        amountsLeft.push(FIRST_PRESALE_AMOUNT * EXTRA_DIGITS_MULTIPLIER);
+        amountsLeft.push(SECOND_PRESALE_AMOUNT * EXTRA_DIGITS_MULTIPLIER);
+        amountsLeft.push(THIRD_PRESALE_AMOUNT * EXTRA_DIGITS_MULTIPLIER);
     }
     
     mapping (address => Balance) usersBalance;
     Transaction[] transactions;
     
-    function buyTokens() public payable {
+    function () external payable {
         require(msg.value >= MINIMUM_PAYMENT_VALUE);
-        uint deposit = msg.value;
+        uint totalDeposit = msg.value;
         
-        uint8 index = 0;
-        uint profit = 0;
+        uint8 stage = 0;
+        uint currentInvestment = 0;
         
         uint totalTokens = 0;
         
-        while (deposit > 0 && index < 3) {
-            uint ethers = 0;
-            uint maxAmount = deposit * MINIMUM_TOKEN_VALUE_MULTIPLIER * prices[index] / 1 ether;
-            if (amounts[index] >= maxAmount) {
-                amounts[index] -= maxAmount;
-                ethers = deposit;
-                deposit = 0;
+        while (totalDeposit > 0 && stage < 3) {
+            uint currentStageDeposit = 0;
+            
+            uint maxTokensAmountForCurrentStage = totalDeposit * amountBoughtInOneEth[stage];
+            //checking free tokens amount for current stage
+            if (amountsLeft[stage] >= maxTokensAmountForCurrentStage) {
+                amountsLeft[stage] -= maxTokensAmountForCurrentStage;
+                currentStageDeposit = totalDeposit;
+                totalDeposit = 0;
             } else {
-                maxAmount = amounts[index];
-                amounts[index] = 0;
-                ethers = maxAmount * 1 ether / MINIMUM_TOKEN_VALUE_MULTIPLIER / prices[index];
-                deposit -= ethers;
+                maxTokensAmountForCurrentStage = amountsLeft[stage];
+                amountsLeft[stage] = 0;
+                currentStageDeposit = maxTokensAmountForCurrentStage / amountBoughtInOneEth[stage];
+                totalDeposit -= currentStageDeposit;

```

Comment:
need to use safe math here. i dont want to accidently have total deposit = 99999999999, and return it to the user.